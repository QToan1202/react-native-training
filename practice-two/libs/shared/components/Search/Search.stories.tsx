import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { Search } from '../index'

export default {
  title: 'components/Search',
  component: Search,
} as ComponentMeta<typeof Search>

const Template: ComponentStory<typeof Search> = (args) => <Search {...args} />

export const Default = Template.bind({})
Default.args = {
  placeholder: 'Search Product',
}
