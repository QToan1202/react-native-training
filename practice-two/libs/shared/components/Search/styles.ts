import { StyleSheet } from 'react-native'

import { COLORS, FONT_FAMILY } from '../../constants'

const styles = StyleSheet.create({
  search: {
    fontFamily: FONT_FAMILY.MONTSERRAT[500],
    fontSize: 14,
    color: COLORS.GRAY_50,
    paddingVertical: 8,
    paddingLeft: 0,
  },
})

export default styles
