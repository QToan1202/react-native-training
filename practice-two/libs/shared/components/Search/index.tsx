import { useCallback, useRef } from 'react'
import { useForm } from 'react-hook-form'
import { TextInput } from 'react-native'
import { Stack, XStack, getTokenValue } from 'tamagui'

import SearchIcon from '@assets/search.svg'

import Input from '../Input'
import { IForm } from '../../types'
import styles from './styles'
import { StyledInputProps } from '../Input/styles'

export type SearchProps = {
  placeholder: string
} & Omit<StyledInputProps, 'label'>

const Search = ({ placeholder, ...rest }: SearchProps) => {
  const { control } = useForm<IForm>()
  const searchInput = useRef<TextInput>(null)

  /**
   * Provide a focus to input when user press on the touchable element
   * User may press on the icon
   * but it will focus in the input element though
   */
  const handlePress = useCallback(() => {
    searchInput.current?.focus()
  }, [])

  return (
    <XStack>
      <XStack
        flex={1}
        alignItems="center"
        space="$space.4"
        paddingHorizontal="$space.4"
        borderRadius="$radius.12"
        backgroundColor="$color.white"
        onPress={handlePress}
      >
        <Stack width="$icon.width" height="$icon.height">
          <SearchIcon fill={getTokenValue('$color.primary')} />
        </Stack>
        <Input
          containerStyle={{ flex: 1 }}
          ref={searchInput}
          name="search"
          control={control}
          style={styles.search}
          placeholder={placeholder}
          placeholderTextColor={getTokenValue('$color.gray_50')}
          {...rest}
        />
      </XStack>
    </XStack>
  )
}

export default Search
