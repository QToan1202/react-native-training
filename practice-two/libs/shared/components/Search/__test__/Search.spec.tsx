import { render, fireEvent } from '@testing-library/react-native'

import Search from '../index'

describe('Testing Search component', () => {
  it('Render Search component success', () => {
    const component = render(<Search placeholder="my search input" />).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('When change input value onChangeText must be called with right props', () => {
    const { getByPlaceholderText } = render(<Search placeholder="my search input" />)
    const element = getByPlaceholderText('my search input')

    fireEvent.changeText(element, 'new products')
    expect(element.props.value).toEqual('new products')
  })
})
