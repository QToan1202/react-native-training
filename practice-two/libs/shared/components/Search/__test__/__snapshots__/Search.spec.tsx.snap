// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Search component Render Search component success 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 50,
        "columnGap": 20,
        "display": "flex",
        "flexDirection": "row",
        "paddingLeft": 17,
      },
      undefined,
    ]
  }
>
  <Image
    source={
      {
        "testUri": "../../../assets/search.png",
      }
    }
    style={
      {
        "height": 24,
        "resizeMode": "cover",
        "width": 24,
      }
    }
  />
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="my search input"
      placeholderTextColor="#4f4f4f"
      style={
        [
          {
            "borderColor": "#fff",
            "borderRadius": 50,
            "borderWidth": 1,
            "color": "#fff",
            "fontFamily": "Montserrat_400Regular",
            "fontSize": 18,
            "paddingLeft": 15,
            "paddingVertical": 12,
          },
          {},
          {
            "color": "#4f4f4f",
            "fontFamily": "Montserrat_500Medium",
            "fontSize": 14,
            "paddingLeft": 0,
            "paddingVertical": 8,
          },
        ]
      }
      value=""
    />
  </View>
</View>
`;
